### PR TITLE
Proposal for fixed sync script

### PR DESCRIPTION
**Pull request by AlCalzone**

### CODE REVIEW

File: temp_OpenHab_sync.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// @ts-check
```
Comment (AlCalzone) [inline temp_OpenHab_sync.js:1]: This is just so VSCode shows me some errors automatically

File: temp_OpenHab_sync.js [hunk 2]
```diff
@@ -18,196 +20,199 @@ var connected;
...
 // Read all enums related to enum.functions
-functions = getEnums('functions');
+/** @type {any[]} */
+const functions = getEnums('functions');
 
 // Loop on all values and store members of value "device" to variable
-for (var i in functions){
-    if (functions[i].name == devices){
-        device = functions[i].members;
+for (const func of functions) {
```
Comment (AlCalzone) [inline temp_OpenHab_sync.js:38]: Prefer `for-of` (items) to `for-in` (indizes)

**Review by AlCalzone — COMMENTED** [review]
I also added a few targeted comments throughout the code.

File: temp_OpenHab_sync.js [hunk 3]
```diff
@@ -18,196 +20,199 @@ var connected;
 var syncronizing = false;
 var device, timeout;
 
+/** 
+ * @type {Map<string, any>}
+ * A map of all expected state changes to prevent infinite loops.
+ * Maps each controlled state to its expected value
+ */
+const expectedStateChanges = new Map();
```
Comment (AlCalzone) [inline temp_OpenHab_sync.js:28]: This remembers which state changes we expect

File: temp_OpenHab_sync.js [hunk 4]
```diff
@@ -18,196 +20,199 @@ var connected;
...
-return connected});
+    return connected
+});
+
+/**
+ * Tests if a state change needs to be synced
+ * @param {string} objID The ID of the state to be checked
+ * @param {any} objVal The value of the state to be checked
+ */
+function isSyncNecessary(objID, objVal) {
```
Comment (AlCalzone) [inline temp_OpenHab_sync.js:135]: This checks if we expected a state change and if its withint the tolerance

File: temp_OpenHab_sync.js [hunk 5]
```diff
@@ -18,196 +20,199 @@ var connected;
...
+    }
+    return true;
+}
+
+/**
+ * Remembers the expected value for a state change
+ * @param {string} objID The ID of the state
+ * @param {any} expectedValue The expected value for the state
+ */
+function rememberExpectedStateChange(objID, expectedValue) {
```
Comment (AlCalzone) [inline temp_OpenHab_sync.js:160]: This remembers which state changes we expect in the near future

**Review by DutchmanNL — APPROVED** [review]

**Comment by DutchmanNL** [comment]
merged

**Pull request merged**